### PR TITLE
Addition of resource and data source for Concrete Interface (vnsCIf)

### PR DESCRIPTION
**Pull request by shrsr**
…rface: vnsCIf

### CODE REVIEW

File: examples/concrete_interface/main.tf [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  vm_name                          = "tenant1-ASA1"
+  relation_vns_rs_c_dev_to_ctrlr_p = "uni/vmmp-VMware/dom-ACI-vDS/ctrlr-vcenter"
+}
+
+resource "aci_concrete_interface" "example" {
+  concrete_device_dn            = aci_concrete_device.concrete.id
+  name                          = "g0/4"
+  encap                         = "unknown"
+  vnic_name                     = "Network adapter 5"
+  relation_vns_rs_c_if_path_att = "topology/pod-1/paths-101/pathep-[eth1/1]"
```
Comment (lhercot) [inline examples/concrete_interface/main.tf:54]: Add an example for a physical device.

File: website/docs/d/concrete_interface.html.markdown [hunk 2]
```diff
@@ -0,0 +1,41 @@
+---
+layout: "aci"
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:2]: missing subcategory

File: website/docs/d/concrete_interface.html.markdown [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+```hcl
+data "aci_concrete_interface" "example" {
+  concrete_device_dn  = aci_concrete_device.example.id
+  name                = "example"
+}
+```
+
+## Argument Reference ##
+
+* `concrete_device_dn` - (Required) Distinguished name of parent Concrete Device object.
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:33]: of the parent

File: website/docs/d/concrete_interface.html.markdown [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+* `name` - (Required) Name of object Concrete Interface.
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:34]: of the Concrete Interface object

File: website/docs/d/concrete_interface.html.markdown [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+## Attribute Reference ##
+* `id` - Attribute id set to the Dn of the Concrete Interface.
+* `name_alias` - (Optional) Name Alias of object Concrete Interface.
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:38]: of the Concrete Interface object

File: website/docs/d/concrete_interface.html.markdown [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+* `encap` - (Optional) The port encapsulation. Type: String.
+* `vnic_name` - (Optional) The concrete interface's vNIC name. Type: String.
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:40]: The virtual NIC (vNIC) name of the L4-L7 Device VM represented by the concrete interface.

File: website/docs/d/concrete_interface.html.markdown [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+* `relation_vns_rs_c_if_path_att` - (Optional) Represents a relation from Concrete Interface to the Physical Port on the Leaf (class fabricPathEp). Note that this relation is an internal object. Type: String.
```
Comment (lhercot) [inline website/docs/d/concrete_interface.html.markdown:41]: from the Concrete Interface

Remove "Note that this relation is an internal object"

File: website/docs/r/concrete_interface.html.markdown [hunk 8]
```diff
@@ -0,0 +1,49 @@
+---
+layout: "aci"
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:2]: missing subcategory

File: website/docs/r/concrete_interface.html.markdown [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+## Example Usage ##
+
+```hcl
+resource "aci_concrete_interface" "example" {
+  concrete_device_dn            = aci_concrete_device.concrete.id
+  name                          = "g0/4"
+  encap                         = "unknown"
+  vnic_name                     = "Network adapter 5"
+  relation_vns_rs_c_if_path_att = "topology/pod-1/paths-101/pathep-[eth1/1]"
+}
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:31]: Add an example for physical interface

File: website/docs/r/concrete_interface.html.markdown [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+## Argument Reference ##
+
+* `concrete_device_dn` - (Required) Distinguished name of the parent Concrete Device object.
+* `name` - (Required) Name of the object Concrete Interface.
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:37]: of the Concrete Interface object

File: website/docs/r/concrete_interface.html.markdown [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+* `encap` - (Optional) The port encapsulation. Type: String.
+* `vnic_name` - (Optional) The concrete interface's vNIC name. Type: String.
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:39]: The virtual NIC (vNIC) name of the L4-L7 Device VM represented by the concrete interface.

File: website/docs/r/concrete_interface.html.markdown [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+* `relation_vns_rs_c_if_path_att` - (Optional) Represents a relation from Concrete Interface to the Physical Port on the Leaf (class fabricPathEp). Note that this relation is an internal object. Type: String.
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:40]: from the Concrete

Remove "Note that this relation is an internal object."

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_vnscif.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+		d.Set("concrete_device_dn", GetParentDn(vnsCIf.DistinguishedName, fmt.Sprintf("/cIf-[%s]", vnsCIfMap["name"])))
+	}
+	d.Set("encap", vnsCIfMap["encap"])
+	d.Set("name", vnsCIfMap["name"])
+	d.Set("vnic_name", vnsCIfMap["vnicName"])
+	d.Set("name_alias", vnsCIfMap["nameAlias"])
+	/* vnsLDevVip does not support annotation due to which we explicitly set annotation to an empty string to override
+	the default value of annotation in the base attributes file. We do this to overcome an inconsistency between imported
+	configuration and a matching configuration waiting to be pushed.*/
+	d.Set("annotation", "")
```
[the remark this answers is not in the document]
Comment (shrsr) [reply aci/resource_aci_vnscif.go]: Took care of this. Annotation was missing in the models file

File: examples/concrete_interface/main.tf [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+# Creating an interface for a Physical Concrete Device
+resource "aci_concrete_interface" "example2" {
+  concrete_device_dn            = aci_concrete_device.concrete.id
+  name                          = "g0/3"
+}
```
Comment (lhercot) [inline examples/concrete_interface/main.tf:62]: Create a separate l4_l7_device and concrete device for this physical one. Just the interface to a virtual device will be more confusing.

File: website/docs/r/concrete_interface.html.markdown [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  relation_vns_rs_c_if_path_att = "topology/pod-1/paths-101/pathep-[eth1/1]"
+}
+```
+
+### Creating an interface for a Physical Concrete Device ###
+
+```hcl
+resource "aci_concrete_interface" "example2" {
+  concrete_device_dn            = aci_concrete_device.concrete.id
+  name                          = "g0/4"
```
Comment (lhercot) [inline website/docs/r/concrete_interface.html.markdown:42]: No relationship to physical interface in this case?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: website/docs/r/concrete_interface.html.markdown (hunk 15 above)
Comment (shrsr) [reply website/docs/r/concrete_interface.html.markdown:42]: Added one

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**